prost-build/src/context.rs

```diff
@@ -0,0 +1,285 @@
+use std::borrow::Cow;
+
+use prost_types::{
+    field_descriptor_proto::{Label, Type},
+    FieldDescriptorProto,
+};
+
+use crate::extern_paths::ExternPaths;
+use crate::message_graph::MessageGraph;
+use crate::{BytesType, Config, MapType, ServiceGenerator};
+
+/// The context providing all the global information needed to generate code.
+/// It also provides a more disciplined access to Config
+/// and its mutable instance of ServiceGenerator.
+pub struct Context<'a> {
+    config: &'a mut Config,
+    message_graph: MessageGraph,
+    extern_paths: ExternPaths,
+}
+
+impl<'a> Context<'a> {
+    pub fn new(
+        config: &'a mut Config,
+        message_graph: MessageGraph,
+        extern_paths: ExternPaths,
+    ) -> Self {
+        Self {
+            config,
+            message_graph,
+            extern_paths,
+        }
+    }
+
+    pub fn config(&self) -> &Config {
+        self.config
+    }
+
+    pub fn service_generator(&mut self) -> Option<&mut (dyn ServiceGenerator + 'static)> {
```

Comment:
... but it can get a reference to the mutable object providing a `ServiceGenerator` vtable. Which was the only reason for passing a mutable reference to `Config` to `CodeGenerator` before, even though it does not need to (and really should not) mutate any other members of `Config`.